src/main/java/com/flipkart/zjsonpatch/JsonPointer.java

```diff
@@ -0,0 +1,313 @@
+                if (!current.has(token.getField()))
+                    error(idx,"Missing field \"" + token.getField() + "\"", document);
+                current = current.get(token.getField());
+            }
+            else
+                error(idx, "Can't reference past scalar value", document);
+        }
+
+        return current;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        JsonPointer that = (JsonPointer) o;
+
+        // Probably incorrect - comparing Object[] arrays with Arrays.equals
+        return Arrays.equals(tokens, that.tokens);
+    }
+
+    @Override
+    public int hashCode() {
+        return Arrays.hashCode(tokens);
+    }
+
+    /** Represents a single JSON Pointer reference token. */
+    static class RefToken {
+        private String decodedToken;
+        transient private Integer index = null;
+
+        public RefToken(String decodedToken) {
+            if (decodedToken == null) throw new IllegalArgumentException("Token can't be null");
+            this.decodedToken = decodedToken;
+        }
+
+        private static final Pattern DECODED_TILDA_PATTERN = Pattern.compile("~0");
+        private static final Pattern DECODED_SLASH_PATTERN = Pattern.compile("~1");
+
+        private static String decodePath(Object object) {
+            String path = object.toString(); // see http://tools.ietf.org/html/rfc6901#section-4
+            path = DECODED_SLASH_PATTERN.matcher(path).replaceAll("/");
+            return DECODED_TILDA_PATTERN.matcher(path).replaceAll("~");
+        }
+
+        private static final Pattern ENCODED_TILDA_PATTERN = Pattern.compile("~");
+        private static final Pattern ENCODED_SLASH_PATTERN = Pattern.compile("/");
+
+        private static String encodePath(Object object) {
+            String path = object.toString(); // see http://tools.ietf.org/html/rfc6901#section-4
+            path = ENCODED_TILDA_PATTERN.matcher(path).replaceAll("~0");
+            return ENCODED_SLASH_PATTERN.matcher(path).replaceAll("~1");
+        }
+
+        private static final Pattern VALID_ARRAY_IND = Pattern.compile("-|0|(?:[1-9][0-9]*)");
+
+        public static RefToken parse(String rawToken) {
+            if (rawToken == null) throw new IllegalArgumentException("Token can't be null");
+            return new RefToken(decodePath(rawToken));
+        }
+
+        public boolean isArrayIndex() {
+            if (index != null) return true;
+            Matcher matcher = VALID_ARRAY_IND.matcher(decodedToken);
+            if (matcher.matches()) {
+                index = matcher.group().equals("-") ? LAST_INDEX : Integer.parseInt(matcher.group());
+                return true;
+            }
+            return false;
+        }
+
+        public int getIndex() {
+            if (!isArrayIndex()) throw new IllegalStateException("Object operation on array target");
```

Comment:
Maybe I misunderstood, but did you mean "Array operation on object target."? You are trying to get an index when a RefToken is not an ArrayIndex

Reply:
Nope, you seem to misunderstand -- a reference token _can_ be a valid array index, but if it's not then it's considered a field reference. Therefore it's an invalid state for `RefToken.getIndex`.